### PR TITLE
[Fix #11984] Add configurable server idle timeout

### LINKED ISSUE #11984: Rubocop server auto-shutdown if idle
## Is your feature request related to a problem? Please describe.

I run Rubocop in git commit hooks, and I the server feature has been a very welcome addition since I no longer have to wait several seconds on every commit attempt.
However, after I'm done with a task I don't like having leftover background processes idling, and it's easy to forget to kill the running Rubocop server after I'm done doing Ruby things for the day.

## Describe the solution you'd like

I would love to have access to a CLI option or environment variable that Rubocop reads when starting a server and which configures the server to automatically shut down if it doesn't receive any requests for a certain amount of time.

For example, `--idle-timeout <seconds>` or `RUBOCOP_SERVER_IDLE_TIMEOUT`.

## Describe alternatives you've considered

I considered writing my own wrapper around Rubocop to handle this use case, but it's impractical to implement from outside of the server process.

### PR DESCRIPTION
**Pull request by Palladinium**
This adds support for a new environment variable when running rubocop in server mode, `RUBOCOP_SERVER_IDLE_TIMEOUT`.

When the variable is set, it's interpreted as a number of seconds, and the server will automatically exit if that many seconds pass without receiving any connections.

I'm hesitant to add tests since they'll likely be timing-sensitive (read: inconsistent) and/or might slow down the suite with `sleep`s.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [ ] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by bbatsov** [comment]
I'll defer to @koic on this one, but I'll mention that it might be nice to be able to set the timeout without with some command-line flag as well. (it would take precedence over the env var)

**Comment by koic** [comment]
> I'm hesitant to add tests since they'll likely be timing-sensitive (read: inconsistent) and/or might slow down the suite with `sleep`s.

It looks good to have a test that uses stubs and mocks. Especially the behavior around the `IO::WaitReadable` exception handling.

File: lib/rubocop/server/core.rb [hunk 1]
```diff
@@ -71,15 +85,18 @@ def server_mode?
...
       end
 
-      def start_server(host, port)
-        @server = TCPServer.open(host, port)
-
+      def output_stream
         # JSON format does not expected output message when IDE integration with server mode.
         # See: https://github.com/rubocop/rubocop/issues/11164
-        return if use_json_format?
+        return nil if use_json_format?
```
Comment (koic) [inline lib/rubocop/server/core.rb:91]: Is this change intentional?

Comment (Palladinium) [reply lib/rubocop/server/core.rb:91]: I think it was just force of habit, I forgot the two are equivalent. I'll revert it since the older line is more concise.

**Comment by bbatsov** [comment]
I'm closing this PR due to no recent activity. Feel free to re-open it if you ever come back to it.
